SeaPublicWebsite/Services/EnergyEfficiency/PropertyDataService.cs

```diff
@@ -13,40 +15,91 @@ public class PropertyDataService(
+
+        var newRecommendations = recommendationsWithPriceCap.Recommendations.Select(r =>
+            new PropertyRecommendation
             {
-                propertyData.EnergyPriceCapYear = null;
-                propertyData.EnergyPriceCapMonthIndex = null;
+                Key = r.Key,
+                Title = r.Title,
+                MinInstallCost = r.MinInstallCost,
+                MaxInstallCost = r.MaxInstallCost,
+                Saving = r.Saving,
+                LifetimeSaving = r.LifetimeSaving,
+                Lifetime = r.Lifetime,
+                Summary = r.Summary
             }
+        ).ToList();
+
+        var recommendationsChanged = RecommendationsChanged(propertyData.PropertyRecommendations, newRecommendations);
+
+        if (recommendationsChanged)
+        {
+            // if no previous recommendations remaining, then this is the users first time through the form
+            // or they have changed their answers (in which case previous answers are discarded anyway).
+            // if so mark this as false to show new visit wording
+            propertyData.RecommendationsUpdatedSinceLastVisit = propertyData.PropertyRecommendations.Count > 0;
+            
+            MergeExistingRecommendationActions(propertyData.PropertyRecommendations, newRecommendations);
+            propertyData.PropertyRecommendations = newRecommendations;
+        }
+        else
+        {
+            propertyData.RecommendationsUpdatedSinceLastVisit = false;
+        }
+        
+        propertyData.RecommendationsLastRetrievedAt = DateTime.UtcNow;
+
+        propertyData.EnergyPriceCapInfoRequested = true;
+
+        if (recommendationsWithPriceCap.EnergyPriceCapInfo is not null)
+        {
+            propertyData.EnergyPriceCapYear = recommendationsWithPriceCap.EnergyPriceCapInfo.Year;
+            propertyData.EnergyPriceCapMonthIndex = recommendationsWithPriceCap.EnergyPriceCapInfo.MonthIndex;
+        }
+        else
+        {
+            propertyData.EnergyPriceCapYear = null;
+            propertyData.EnergyPriceCapMonthIndex = null;
         }
 
         propertyData.HasSeenRecommendations = true;
         await propertyDataStore.SavePropertyDataAsync(propertyData);
         return propertyData;
     }
+
+    private bool RecommendationsChanged(List<PropertyRecommendation> existingRecommendations,
+        List<PropertyRecommendation> newRecommendations)
+    {
+        if (existingRecommendations is null || existingRecommendations.Count == 0)
+        {
+            return true;
+        }
+
+        if (existingRecommendations.Count != newRecommendations.Count)
+        {
+            return true;
+        }
+
+        return existingRecommendations.Any(er =>
+            !newRecommendations.Any(nr => nr.HasSameImpactAs(er)));
+    }
+
+    private void MergeExistingRecommendationActions(List<PropertyRecommendation> existingRecommendations,
```

Comment:
happy to hear different ways on structuring this. ideally would not mutate at all but thought this could be tricky to structure

Reply:
left as blank. in this case it'd be that we've now received a new recommendation so they need to decide on it

Reply:
I was going to suggest a change here to make this function a more-usual merge, than an update of the new recs with the old recs (e.g. return a list, not update the new recs). But am I correct in thinking that we want to have the new titles / summaries from the new recs always, and not take the old recommendations as they are in any case? In which case I think this does work -
I think lets rename this to something like "updateRecommendationsWithPriorActions" or similar (I can't seem to get the character count down there any more 😅). It makes it clear this updates the list

Reply:
have renamed this and squashed. needs approval again